src/Model/Behavior/ProfferBehavior.php

```diff
@@ -79,53 +79,131 @@ public function beforeMarshal(Event $event, ArrayObject $data, ArrayObject $opti
-                    if (getimagesize($path->fullPath()) !== false && isset($settings['thumbnailSizes'])) {
-                        // Allow the transformation class to be injected
-                        if (!empty($settings['transformClass'])) {
-                            $imageTransform = new $settings['transformClass']($this->_table, $path);
-                        } else {
-                            $imageTransform = new ImageTransform($this->_table, $path);
-                        }
+        if ($tableEntityClass !== null && $entity instanceof $tableEntityClass) {
+            $uploadList = [
+                [
+                    'name' => $entity->get('name'),
+                    'type' => $entity->get('type'),
+                    'tmp_name' => $entity->get('tmp_name'),
+                    'error' => $entity->get('error'),
+                    'size' => $entity->get('size')
+                ]
+            ];
+        } else {
+            $uploadList = $entity->get($field);
+            if (count(array_filter(array_keys($entity->get($field)), 'is_string')) > 0) {
+                $uploadList = [$entity->get($field)];
+            }
+        }
 
-                        $thumbnailPaths = $imageTransform->processThumbnails($settings);
-                        $imagePaths = array_merge($imagePaths, $thumbnailPaths);
+        foreach ($uploadList as $upload) {
+            if ($this->moveUploadedFile($upload['tmp_name'], $path->fullPath())) {
+                $entity->set($field, $path->getFilename());
+                $entity->set($settings['dir'], $path->getSeed());
 
-                        $eventData = ['path' => $path, 'images' => $imagePaths];
-                        $event = new Event('Proffer.afterCreateImage', $entity, $eventData);
-                        $this->_table->eventManager()->dispatch($event);
-                    }
-                } else {
-                    throw new Exception('Cannot upload file');
-                }
+                $this->createThumbnails($entity, $settings, $path);
+            } else {
+                throw new Exception('Cannot upload file');
             }
-            unset($path);
         }
 
-        return true;
+        unset($path);
+    }
+
+    /**
+     * Load a path class instance and create the path for the uploads to be moved into
+     *
+     * @param \Cake\Datasource\EntityInterface $entity Instance of the entity
+     * @param string $field The upload field name
+     * @param array $settings Array of upload settings for the field
+     * @param \Proffer\Lib\ProfferPathInterface|null $path Inject an instance of ProfferPath
+     *
+     * @return \Proffer\Lib\ProfferPathInterface
+     */
+    protected function createPath(EntityInterface $entity, $field, array $settings, ProfferPathInterface $path = null)
+    {
+        if (!empty($settings['pathClass'])) {
+            $path = new $settings['pathClass']($this->_table, $entity, $field, $settings);
```

Comment:
They're good points, but I'm not sure how else to approach this. The premise of the feature is to allow userland code to overload some of the core components so that they can implement their own logic. I guess I could check if it implements the interface.